pkg/container/versioned/value.go

```diff
@@ -0,0 +1,425 @@
+// are available for use for as long as the VersionHold exists.
+//
+// A special form with a nil manager, which is returned by Latest(), always finds the latest
+// versions and does not keep any versions from being released.
+type VersionHold struct {
+	// 'version' is the version this VersionHold keeps from being released
+	version version
+	// Manager of this VersionHold, if any. Used for closing if non-nil.
+	// Atomic due to nilling and for copy prevention.
+	manager atomic.Pointer[Manager]
+}
+
+// versionHoldFinalizer is used to warn about missing Close() calls.
+func versionHoldFinalizer(h *VersionHold) {
+	log.Errorf("Hold for SelectorCache version %d not closed.", h.version)
+	h.Close()
+}
+
+func newVersionHold(version version, manager *Manager) *VersionHold {
+	// hold on maxVersion never expires
+	if version == maxVersion {
+		manager = nil
+	}
+	h := &VersionHold{version: version}
+	h.manager.Store(manager)
+	if manager != nil {
+		// Set a finalizer to catch unclosed holds. The finalizer function complains
+		// loudly, so that we do not rely the finalizer for closing.
+		runtime.SetFinalizer(h, versionHoldFinalizer)
+	}
+	return h
+}
+
+// Latest returns a VersionHold for the latest version of current/non-removed values
+// Only to be used in cases where the latest values are needed and no transactionality is required.
+func Latest() *VersionHold {
+	return newVersionHold(maxVersion, nil)
+}
+
+func (h *VersionHold) IsValid() bool {
+	return h != nil && (h.version == maxVersion || h.manager.Load() != nil)
+}
+
+func (h *VersionHold) Close() (err error) {
+	if !h.IsValid() {
+		return errInvalidVersion
+	}
+	manager := h.manager.Load()
+	if manager != nil {
+		runtime.SetFinalizer(h, nil)
+		err = manager.releaseVersion(h.version)
+		h.manager.Store(nil)
+	}
+	return err
+}
+
+type atomicVersion struct {
+	version atomic.Uint64
+}
+
+func (a *atomicVersion) load() version {
+	return version(a.version.Load())
+}
+
+func (a *atomicVersion) store(version version) {
+	a.version.Store(uint64(version))
+}
+
+func (a *atomicVersion) compareAndSwap(old, new version) bool {
+	return a.version.CompareAndSwap(uint64(old), uint64(new))
+}
+
+type versionCount struct {
+	version version
+	count   int
+}
+
+// Manager defines a commont version number space for multiple versioned.Values,
+// and provides facilities for cleaning out stale versions.
+type Manager struct {
+	version atomicVersion
+
+	// cleaner is called with the earliest version that must be kept
+	cleaner func(KeepVersion)
+
+	// mutex protects the rest of the fields
+	mutex lock.RWMutex
+
+	// versions is an ordered list of outstanding VersionHolds with a reference count.
+	// Outdated values can be cleaned when there are no outstanding VersionHolds for them.
+	versions []versionCount
+
+	// latest cleaned version
+	cleanedVersion version
+}
+
+// NewVersionManager returns a new VersionManager
```

Comment:
Match this with `Manager`? e.g. `NewManager`. Using `New` would be more go-like, but that's probably confusing.

Reply:
Changed to `NewManager`. When there are multiple types in a package a bare `New` would indeed be confusing.